muzlag/internal/service/player/play.go

```diff
@@ -0,0 +1,151 @@
+package player
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+
+	"github.com/bwmarrin/discordgo"
+	"layeh.com/gopus"
+
+	"github.com/Drozd0f/gobots/muzlag/pkg/ytdl"
+)
+
+const (
+	frameSize int = 960           // uint16 size of each audio frame
+	maxBytes      = frameSize * 4 // max size of opus data
+)
+
+func (s *ServicePlayer) Play(vc *discordgo.VoiceConnection, url string) error {
+	stream, err := s.dl.GetAudioStream(url, ytdl.BestAudioFormat)
+	if err != nil {
+		return fmt.Errorf("ytdl get audio stream: %w", err)
+	}
+
+	defer func() {
+		if err = stream.Cancel(); err != nil {
+			s.logger.Error("stream cancel", slog.Any("error", err))
+		}
+	}()
+
+	sout, err := stream.GetOutput()
+	if err != nil {
+		return fmt.Errorf("ytdl stdout pipe: %w", err)
+	}
+
+	defer func() {
+		if err = sout.Close(); err != nil {
+			s.logger.Error("stream out close", slog.Any("error", err))
+		}
+	}()
+
+	player := s.ffmpeg.PlayerFromInput(sout)
+	defer func() {
+		if err = player.Cancel(); err != nil {
+			s.logger.Error("player close", slog.Any("error", err))
+		}
+	}()
+
+	pout, err := player.GetOutput()
+	if err != nil {
+		return fmt.Errorf("player get output: %w", err)
+	}
+
+	defer func() {
+		if err = pout.Close(); err != nil {
+			s.logger.Error("player output close", slog.Any("error", err))
+		}
+	}()
+
+	if err = stream.Start(); err != nil {
+		return fmt.Errorf("stream start: %w", err)
+	}
+
+	// Starts the ffmpeg command
+	if err = player.Start(); err != nil {
+		return fmt.Errorf("player start: %w", err)
+	}
+
+	// Send "speaking" packet over the voice websocket
+	if err = vc.Speaking(true); err != nil {
+		return fmt.Errorf("voice connection speaking: %w", err)
+	}
+
+	// Send not "speaking" packet over the websocket when we finish
+	defer func() {
+		if err = vc.Speaking(false); err != nil {
+			s.logger.Error("voice connection speaking false", slog.Any("error", err))
+		}
+	}()
+
+	send := make(chan []int16, 2)
+	defer close(send)
+
+	done := make(chan bool)
+	go func() {
+		s.sendPCM(vc, send)
+		done <- true
+	}()
+
+	for {
+		// read data from ffmpeg stdout
+		audiobuf := make([]int16, frameSize*2)
+
+		err = binary.Read(pout, binary.LittleEndian, &audiobuf)
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
```

Comment:
io.EOF means, that you read last peace of data, but you are not handling it here. Not critical, but you'll lose chunk of data probably

Reply:
It is interesting that there were no losses frame when processing this error, if this happens in the future I will fix this section of the code